File: 23-merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode head = new ListNode(-1);
        ListNode ans = head;
        boolean cont = true;
        int least = 0;
        while(cont){
            int min = Integer.MAX_VALUE;
            for (int i = 0;i<lists.length;i++){
                ListNode h = lists[i];
                if (h != null){
                    if (min > h.val){
                        min = h.val;
                        least = i;
                    }  
                }
                
            }

            if (min!=Integer.MAX_VALUE){
                head.next = new ListNode(min);
                head = head.next;
                lists[least] = lists[least].next;
            }
            else{
                cont = false;
            }
            

        }

        return ans.next;

    }
}

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class HeapNode:
    def __init__(self, node):
        self.node = node

    def __lt__(self, other):
        return self.node.val < other.node.val

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:

        pq = []
        ans = []
        x = len(lists)
        for i in range(x):
            if lists[i]:
                heapq.heappush(pq, HeapNode(lists[i]))

        while pq:
            heapnode = heapq.heappop(pq)
            ans.append(heapnode.node.val)

            if  heapnode.node.next is not None:
                heapq.heappush(pq, HeapNode(heapnode.node.next))

        
        dummy = ListNode()
        head = dummy
        for num in ans:
            dummy.next = ListNode(num)
            dummy = dummy.next

        return head.next
